### PR TITLE
FAT: Add support for long file names

### PR DESCRIPTION
**Pull request by Deryugin**

### CODE REVIEW

**Comment by alexkalmuk** [comment]
Can you please provide more information about how we can check this?

**Comment by Deryugin** [comment]
> Can you please provide more information about how we can check this?

Basic steps are

dd if=/dev/zero of=usbdisk.img count=65536 bs=2048
mkfs.vfat -n "LINUX" usbdisk.img -I

./scripts/qemu/auto_qemu -hda usbdisk.img
in Embox:
mkdir -v /mnt
mount -t vfat /dev/hda /mnt
echo 123 > /mnt/longlowcasename.qwerty

quit Qemu

sudo mount usbdisk.img /mnt
ls /mnt # Should show file with a long name

**Comment by AntonKozlov** [comment]
A new test in test/fs? :)

**Comment by AntonKozlov** [comment]
Sorry! wrong button

**Comment by alexkalmuk** [comment]
> > Can you please provide more information about how we can check this?
> 
> Basic steps are
> 
> dd if=/dev/zero of=usbdisk.img count=65536 bs=2048
> mkfs.vfat -n "LINUX" usbdisk.img -I
> 
> Build x86/qemu template
> ./scripts/qemu/auto_qemu -hda usbdisk.img
> in Embox:
> mkdir -v /mnt
> mount -t vfat /dev/hda /mnt
> echo 123 > /mnt/longlowcasename.qwerty
> 
> quit Qemu
> 
> sudo mount usbdisk.img /mnt
> ls /mnt # Should show file with a long name

root@embox:/#echo 123 > /mnt/longlowcasename.qwerty
echo: Command returned with code -60: File name too long
tish error: #196

**Comment by Deryugin** [comment]
@alexkalmuk try new VFS, please

**Comment by alexkalmuk** [comment]
@Deryugin Can you please attach here your mods.config? I got assertion failure on `>mount -t binfs / /bin` with new vfs.

**Comment by Deryugin** [comment]
New VFS has no binfs driver..

I pushed fixes for old VFS, it should work fine now.

**Comment by alexkalmuk** [comment]
@Deryugin OK, I removed binfs from start script and your test passed. Thanks!

But this one doesn't work for me:

```
root@embox:/#mkdir -v /mnt                                                      
root@embox:/#mount -t vfat /dev/hda /mnt                                        
root@embox:/#echo 123 > /mnt/longlowcasename.qwerty
root@embox:/#ls /mnt
 /mnt/longlowcasename.qwerty
root@embox:/#echo 123 > /mnt/longlonglowcasename.qwerty
root@embox:/#ls /mnt
 /mnt/longlowcasename.qwerty
root@embox:/#
```

I expected to see both files created.

File: src/fs/driver/fat/fat_common.c [hunk 1]
```diff
@@ -1377,26 +1470,42 @@ uint32_t fat_write_file(struct fat_file_info *fi, uint8_t *p_scratch,
...
+		strncpy(tmppath, (char *) path, filename - (char *) path);
+		filename++;
+	}
+
 	fsi = fi->fsi;
 	volinfo = &fsi->vi;
 
 	fi->mode = mode;
 
+	/*
```
Comment (anton-bondarev) [inline src/fs/driver/fat/fat_common.c]: Rm Or it may be used in the future?

**Pull request merged**